### ISSUE TITLE
Bootstrap-tables issues & fixes

### ISSUE THREAD

**Issue by odiea** [jekkos/opensourcepos#507]
When extracting zip Customer_lang already exists.
Items stock location filter appears not to work.
Item kits does not generate barcodes.
Takings date filtering does not work.

**Comment by jekkos** [comment]
Thanks for creating this new issue. I might want to close the old one as that thread has become too long.

**Comment by odiea** [comment]
Yes I thought so too. So I will report my errors on this thread now.

**Comment by odiea** [comment]
Items. after updating an item then try to update another item again modal form disappears.
Takings. I finally got a receipt to show. In the table view there is no
receipt reprint. Just update and invoice. 
Let me know when these are fixed and I will continue testing.

**Comment by odiea** [comment]
It appears that all the modal forms disappear after doing an update in any module.

**Comment by jekkos** [comment]
@odiea I've fixed most of these issues in my latest commit. What do you mean exactly with customer_lang already present? Also takings date filtering works fine here.. 

Or did you mean that the filter only allows date range?

**Comment by daN4cat** [comment]
sorry I needed to rename this because I scratched my head a bit to connect Bstables to bootstrap tables....

**Comment by odiea** [comment]
When extracting zip Customer_lang already exists. Is the error I receive when extracting the zip into a normal layout.

**Comment by odiea** [comment]
The takings filter only shows pos for the current date. When you try to get a full months worth nothing shows.
In item kits the table does not automatically refresh to show the new item.

**Comment by odiea** [comment]
Item kits starting updating the table after adding a few more items but date filtering in takings is not working properly.

**Comment by odiea** [comment]
When I go to takings after entering a sale nothing shows until I set the Date to today and then after that changing the Date just shows todays Sales. Starting with the default date the Date filter shows nothing in Sales.

**Comment by odiea** [comment]
Here is a screen shot of what I am talking about. This is the default view. It should show a couple hundred Sales.
![2016-04-27 14_21_37-christmas sweater store -- powered by os point of sale](https://cloud.githubusercontent.com/assets/10962177/14866987/8b1740dc-0c83-11e6-8b71-f13381bbb42e.png)

**Comment by jekkos** [comment]
ok so that's the default load state of this page? In my version all seems to work correctly, don't really understand what's going on

**Comment by jekkos** [comment]
![screenshot from 2016-04-27 22 35 08](https://cloud.githubusercontent.com/assets/576234/14867386/87d2e562-0cc8-11e6-8b7f-8e8932bd517e.png)
![screenshot from 2016-04-27 22 35 20](https://cloud.githubusercontent.com/assets/576234/14867387/87d63e42-0cc8-11e6-89d7-646ad3f8a4de.png)
![screenshot from 2016-04-27 22 35 28](https://cloud.githubusercontent.com/assets/576234/14867388/87d98ba6-0cc8-11e6-82cc-73c94606d86f.png)

**Comment by jekkos** [comment]
check the date ranges in my screencaps together with the ones on the sale.. seeems to be all correct? Just tried in chrome and same thing there

**Comment by odiea** [comment]
Here is my view I should have at least 7 or 8 sales showing.
![2016-04-27 14_40_37-christmas sweater store -- powered by os point of sale](https://cloud.githubusercontent.com/assets/10962177/14867513/1c28cbf2-0c86-11e6-9240-ee9ae096ddd5.png)

**Comment by daN4cat** [comment]
Search box looks big.... missing the -sm in the class?

**Comment by odiea** [comment]
I will check later. This is the latest download.

**Comment by odiea** [comment]
@daN4cat if you compare the master manage with boot strap tables manage you can see that the manage in boot strap tables appears not to include the search feature.

**Comment by jekkos** [comment]
Yes that's normal.. search field is from bootstrap-tables code now. It's a plugin builtin. We could alter some css to make it smaller tho (or add a class using js). I'll look into that tomorrow.

Also regarding sales there might be an error in the search query, will look into that a bit further tormorrow as well

**Comment by odiea** [comment]
OK. Thanks for the update.

**Comment by jekkos** [comment]
Ok icon size should be good now.

**Comment by odiea** [comment]
In Items, Bulk update only appears to update the first item and no other. Please check this out as I am on a different computer than my normal one. I did do a fresh download about 20 min ago. I am still having the same issues with Takings as I did on the other computer.

**Comment by odiea** [comment]
I retried on my normal computer and Bulk update is just updating the first item in the table.

**Comment by odiea** [comment]
Also after doing a bulk update that says it was successful and all check marks are removed you select barcodes and all items will create barcodes with no check mark in them and Delete ,Bulk Update and Generate are no longer prevented from being used.

**Comment by jekkos** [comment]
Fixed bulk update refresh .. but I have to look into this last issue you describe here. Would you expect to keep the selected state of the items you have edited or not?

**Comment by daN4cat** [comment]
With regards to the search box size, can this be set to -sm as the rest?

**Comment by jekkos** [comment]
I made the size smaller so should be good.

**Comment by jekkos** [comment]
check https://github.com/jekkos/opensourcepos/commit/c4c2a81753eb4bf9e275c9694891fa15427bf9c4

**Comment by odiea** [comment]
Bulk update does not work with tax% update. It just changes the first row. Also all check boxes are cleared after an update is finished. When I manually add  check mark then remove it the buttons then become unavailable. Refreshing items does eliminate this error.

After I did the update then tried barcode I received this error. 
A PHP Error was encountered
Severity: Warning
Message: Missing argument 1 for Items::generate_barcodes()
Filename: controllers/Items.php
Line Number: 240
Backtrace:
File: C:\xampp\htdocs\bstables\application\controllers\Items.php
Line: 240
Function: _error_handler

File: C:\xampp\htdocs\bstables\index.php
Line: 315
Function: require_once
A PHP Error was encountered
Severity: Notice
Message: Undefined variable: item_ids
Filename: controllers/Items.php
Line Number: 245
Backtrace:
File: C:\xampp\htdocs\bstables\application\controllers\Items.php
Line: 245
Function: _error_handler
File: C:\xampp\htdocs\bstables\index.php
Line: 315
Function: require_once

**Comment by jekkos** [comment]
So left for me to check

* item kits update
* sale time filtering
* bulk update with tax percentage
* checkbox state after form submit (this seems to be a bug in bootstrap tables)

**Comment by odiea** [comment]
Item kit update is not refreshing the tables but adding a new Kit does. Also you can see there are no $ in ft of the Price  numbers. 
![2016-04-28 14_13_30-christmas sweater store -- powered by os point of sale](https://cloud.githubusercontent.com/assets/10962177/14899896/72889cf0-0d4b-11e6-9f67-dfae5a043f7c.png)

**Comment by odiea** [comment]
Let me know when you are finished and I will re download and try some more.

**Comment by odiea** [comment]
Customer tables also do not refresh after an update.

**Comment by odiea** [comment]
So after last pull bulk update in Items does not work. The form shows, the warning shows but nothing happens.

**Comment by jekkos** [comment]
Another day, another bootstrap tables commit 

* Table delete now show animation before removing the selected row
* Item Bulk update should be fine now, including tax update
* Table row selection state should be maintained after dialog close
* Fix for pagination in takings overview (like to have that one retested)
* Item kits are now added / updated properly

**Comment by odiea** [comment]
Bulk update is working now but in Takings Invoice number is not showing.
![2016-05-02 12_16_42-christmas sweater store -- powered by os point of sale](https://cloud.githubusercontent.com/assets/10962177/14963247/f6e96c42-105f-11e6-8fbd-4cae2b02192f.png)

**Comment by odiea** [comment]
Payment type is also not showing.

**Comment by jekkos** [comment]
Ok beside that do you see the expected number of sales on screen?

**Comment by jekkos** [comment]
Ok both issues should be fixed now in latest commit

**Comment by odiea** [comment]
Yes all appears to be good now. Date filtering is also working good.

**Comment by jekkos** [comment]
I've synced with latest version in master and did some testing, looks all good now.

Everyone ok to merge to master? Think this will speed up the bug hunt as more people will start using it.  I will continue rework of reports with the tables later on (I've disabled sale edit in the detailed report for now, not sure if we should keep it?)

**Comment by daN4cat** [comment]
It's probably better to have reports done as well before merging otherwise we will get people posting issues about those straight away.

This time I will give a try and test the latest, seeing odiea reports on bugs and the fact that I was busy with SMS stuff I drop the priority of the new table testing. But now it makes sense to give a boost and merge it.

BTW, why bug & epic? An Epic cannot be a bug, it's a non-sense (if we follow proper Scaled Agile definition).

**Comment by jekkos** [comment]
Was looking for a way to bundle those bugs in this issue but too bad the zenhub thing only allows to see content if the plugin is installed. 

Ok I guess I'll start to look into cleanup of report tabular views then. Think we can improve that part greatly with client side pagination and configurable columns. Table headers and data should be loaded just once and ajax can be used if a detailed row is edited (for which support is present in table_support)

I also haven't found the time yet to add any extra table plugins like the ones that allow csv export, inline edit or responsive display. Think this should be a matter of adding a new js script to the build.

**Comment by RamkrishnaMondal** [comment]
wow! looks great, also wonderfully  responsive. Found one small TBD

![screenshot from 2016-05-03 20 10 23](https://cloud.githubusercontent.com/assets/14151032/14986932/880501ee-116b-11e6-9801-525e09156430.png)

**Comment by odiea** [comment]
With latest download glyphicon-list-alt does not show
![2016-05-03 08_46_23-christmas sweater store -- powered by os point of sale](https://cloud.githubusercontent.com/assets/10962177/14987011/8a6326ec-110b-11e6-8a12-1de6c8a3f4c4.png)

**Comment by RamkrishnaMondal** [comment]
In customers 'selection of number of customers' to display (ie 10, 25. 50, 100) not working.

**Comment by jekkos** [comment]
Thanks for testing. All these issues should be fixed now in latest commit.

**Comment by jekkos** [comment]
Another fix to set default number of rows according to ospos' config setting

**Comment by daN4cat** [comment]
@jekkos if you feel most is done and working about right, then you are correct in your suggestion to merge to master so we avoid further splits. I was just under the impression that reports were not working and worried about the people adding bug reports.

Sorry if I cannot be of much help, I promised to test but I've been caught up on a work load peak and other personal matters.
Also the sun is shining now and it's a kind of pre-summer in London. Hard to work on this stuff after work.

**Comment by jekkos** [comment]
No worries I won't be having much time coming day neither, will be gone cycling for three days. I have started tables conversion in reports anyway and it seems to progress well. I finished all the summary reports and now working on the tabular details with nested table. Once this is finished I might just merge it all at once.

**Comment by daN4cat** [comment]
Just started to use the new tables.

Overall looks good, well done :+1: 

Some comments:

- sorting doesn't seem possible, is that something that needs to be enabled or am I missing something?

**Comment by RamkrishnaMondal** [comment]
@jekkos   Is it possible to change the color of the 'table headers' or table background? If yes where to do that?

**Comment by RamkrishnaMondal** [comment]
Print button on sales Takings page not printing the table.

**Comment by odiea** [comment]
If you are using jsprint have you tried reselecting the printers under receipt config. I have had to do that after each update I have done.

**Comment by RamkrishnaMondal** [comment]
@odiea  You are right, it's working now. But printing one almost blank page as page no 1  if i select 25 or more rows to display.
![screenshot from 2016-05-05 18 37 24](https://cloud.githubusercontent.com/assets/14151032/15044192/cf87c8f4-12f0-11e6-99b3-61492157d355.png)

**Comment by odiea** [comment]
Ok a new bug to look at. It appears this branch isn't quite ready for primetime.

**Comment by daN4cat** [comment]
I fixed the glyphicons part, added a space and completed the glyphincos changes in the buttons

**Comment by daN4cat** [comment]
I noticed Takings starts b default listing all the transactions. I'll change to be Today as default.

**Comment by RamkrishnaMondal** [comment]
For me it is showing last ten transaction , irrespective of date as i have  saved 10 rows to show in config. I think, jekkos has intentionally done that, as he mentioned this on commit description. Better not to change without his consent.

**Comment by daN4cat** [comment]
Original Takings was today only and so it stays.

**Comment by jekkos** [comment]
@RamkrishnaMondal No prob I copied the code for date filtering from items so it wasn't changed intentionally. Just a little easier for testing.

**Comment by daN4cat** [comment]
I removed the delete button from the print in Takings, but the first page is blank and the table is printed in the next pages. Also the table is printed with scroll bars and all the control. There is work to do to get it back to the original idea / layout.

**Comment by jekkos** [comment]
Ok that's something we'll need to  look into. I've committed changes for reports that are all using bs tables as well now.

Detailed sale edit was reworked a bit so now it's consistent with takings. In receivings I temporarily left out the delete buttron as it isn't working at this moment. Have only tested sale update in detailed which is fine. 

Also sorting will need to be implemented in all controllers. Just a matter of passing sortable: true to the bs table constructor. However the backend should add the query parameter to the search for all controllers.

**Comment by odiea** [comment]
Things are looking good. It appears customer search in Sales has quit working and item kits table dollar amounts  does not show the dollar sign $. Report tables look good.

**Comment by daN4cat** [comment]
I checked a bit the reports. The detailed sales looks good, however if you click on the edit icon it doesn't open the dialog box but switches to a view. Probably it's missing the modal-dlg class or similar.

**Comment by daN4cat** [comment]
Same applies to detailed receivings

**Comment by daN4cat** [comment]
Also I have noticed that in Items the bulk edit dialog is missing the title.

**Comment by daN4cat** [comment]
Fixed item kits missing currency symbol

**Comment by daN4cat** [comment]
Fixed the missing bulk edit dialog box title

**Comment by daN4cat** [comment]
Minor fix to item kits delete confirmation dialog box

**Comment by daN4cat** [comment]
Minor fix to detailed sales report as it was using a wrong index.

**Comment by odiea** [comment]
Sales Register - Customer search not working. After adding a letter or name then no drop down shows then pressing enter add new customer form shows.
Return Sales putting in a pos number and clicking pos number shows and you can select it. If you press enter to select the number add new item form shows.

**Comment by RamkrishnaMondal** [comment]
Open detailed sale report, click on the bottom pagination to go to any other page, now  try to edit a sale..... You can't.

**Comment by odiea** [comment]
It appears to loose the modal setting. It looks like it's time to start  per module issues. What do you think? This issue is getting quite long now.

**Comment by RamkrishnaMondal** [comment]
@daN4cat  One good thing i have found on this branch is "Loading, Please wait ......". We also need that on sending email and sms, as these two also take some time. 
Could you please implement "Sending, Please wait ......"  during email and sms sending ?

**Comment by daN4cat** [comment]
The customer suggest in sales/register is broken because of a mistake I made. I'm fixing.

**Comment by daN4cat** [comment]
The sale edit loosing model is something I noticed but then thought it was my installation.
Now retesting and indeed I can reproduce it. That's the correct sequence.
In general there is an issue with modal dialog, I got a couple of time the modal dialog box opening and showing a normal view inside. But I haven't found yet the right sequence to reproduce it.

**Comment by daN4cat** [comment]
Now I found the issue I was talking about.

- go to Items
- list a lot of items.
- click at the bottom to limit to 10
- selected all the items clicking on the top checkbox
- click bulk edit button and you see this:

![image](https://cloud.githubusercontent.com/assets/636048/15094164/4e9921e8-1494-11e6-9e6f-a343d6a98272.png)

**Comment by daN4cat** [comment]
Also note that behind the modal there are two search boxes on the right.

**Comment by daN4cat** [comment]
And now I noticed that empty tax field is printed as false in the list.

**Comment by odiea** [comment]
I cannot reproduce this issue. In Items I select 100 . I then select 10 and try bulk edit. It is working fine for me.

**Comment by odiea** [comment]
It is easy to reproduce in Reports detailed sale then edit after you change the row numbers.

**Comment by odiea** [comment]
Customer search in Sales is working fine now. Thanks

**Comment by jekkos** [comment]
I've done a commit that enable sort in items, customers, suppliers, employees and item kits.

**Comment by daN4cat** [comment]
@jekkos I tried removing table_sorter js plug in but it looks to be still in use. Tables didn't show at all after having removed it. Any reason for the dependency?

**Comment by jekkos** [comment]
There shouldn't be any dependencies indeed.. Could you check your js console and post any error messages here? It's prolly an error that breaks table initialisation.

**Comment by jekkos** [comment]
There might be another issue in that latest. Guess i need to change the uniqueId of the tables after I changed it in the table_helpers for sorting

**Comment by odiea** [comment]
I have Date picker confusion. This Month to Today Last Year. Should it contain this year and last year dates or just last years dates?

**Comment by daN4cat** [comment]
The meaning is to compare the same period of time (current month to today) by in the previous year.
Maybe the English definition is not totally correct, we can change for a better suggestion.

**Comment by RamkrishnaMondal** [comment]
Open detailed sale report, by clicking on the  pagination select any page other than page-1, now try to edit a sale..... You can't.

Just downloaded and tested the latest version, this bug is still there.

**Comment by RamkrishnaMondal** [comment]
@daN4cat 
 Just out of curiosity, what will happen if i select "today last year' on 29th February of 2020?

**Comment by RamkrishnaMondal** [comment]
There is a typo on item's excel import dialogue, it says "import customers from excel sheet".

**Comment by RamkrishnaMondal** [comment]
Print button on Takings  prints a blank first page, if no of rows are more than 10 .... not yet fixed.

**Comment by jekkos** [comment]
@RamkrishnaMondal Indeed the modals need to be reinitialized after loading another page. As the reports do it purely in js it isn't foreseen yet, so I'll need to find a fix.

**Comment by odiea** [comment]
If possible it might say This Month Last Year to Today Last Year

**Comment by jekkos** [comment]
@RamkrishnaMondal @odiea Did fixes for

* Takings print . All columns should be on screen now
* Item search in sale and receiving register
* submit action when enter is hit in search bar (this showed the new item dialog before)

**Comment by jekkos** [comment]
Don't know if there are any bugs left now? Maybe a little more tweaking in the takings print and guess we're good to go for a merge in master?

**Comment by daN4cat** [comment]
Takings print still prints the first page blank and then the report in page 2 and etc,

**Comment by RamkrishnaMondal** [comment]
When everything is getting the touch of  bootstrap, why not report listing?  
![screenshot from 2016-05-14 00 12 16](https://cloud.githubusercontent.com/assets/14151032/15258411/16ee181e-1969-11e6-88ca-35a5e75c12fa.png)

**Comment by daN4cat** [comment]
I believe it's better to close this and be now more specific creating specific issues otherwise all the "catch-all" become a container of all the possible things.

At the same time I converted the original issue driving the introduction of bootstrap-tables to an Epic and link all the issue to that one for future references. I will try to work in the background to group past items in that ways so we cluster topics together under one parent.
